text/0000-ptr-meta.md

```diff
@@ -0,0 +1,387 @@
+
+The bounds on `null()` and `null_mut()` function in that same module
+as well as the `NonNull::dangling` constructor
+are changed from (implicit) `T: Sized` to `T: ?Sized + Thin`.
+Similarly for the `U` type parameter of the `NonNull::cast` method.
+This enables using those functions with [extern types].
+
+The `Pointee` trait is implemented for all types.
+This can be relied on in generic code,
+even if a type parameter `T` does not have an explicit `T: Pointee` bound.
+This is similar to how the `Any` trait can be used without an explicit `T: Any` bound,
+only `T: 'static`, because a blanket `impl<T: 'static> Any for T {…}` exists.
+(Except that `Pointee` is not restricted to `'static`.)
+
+For the purpose of pointer casts being allowed by the `as` operator,
+a pointer to `T` is considered to be thin if `T: Thin` instead of `T: Sized`.
+This similarly includes extern types.
+
+`std::raw::TraitObject` and `std::raw` are deprecated and eventually removed.
+
+[trait alias]: https://github.com/rust-lang/rust/issues/41517
+[extern types]: https://github.com/rust-lang/rust/issues/43467
+
+```rust
+/// This trait is automatically implement for every type.
+///
+/// Raw pointer types and reference types in Rust can be thought of as made of two parts:
+/// a data pointer that contains the memory address of the value, and some metadata.
+///
+/// For statically-sized types (that implement the `Sized` traits)
+/// as well as for `extern` types,
+/// pointers are said to be “thin”: metadata is zero-sized and its type is `()`.
+///
+/// Pointers to [dynamically-sized types][dst] are said to be “fat”
+/// and have non-zero-sized metadata:
+///
+/// * For structs whose last field is a DST, metadata is the metadata for the last field
+/// * For the `str` type, metadata is the length in bytes as `usize`
+/// * For slice types like `[T]`, metadata is the length in items as `usize`
+/// * For trait objects like `dyn SomeTrait`, metadata is [`DynMetadata`].
+///
+/// In the future, the Rust language may gain new kinds of types
+/// that have different pointer metadata.
+///
+/// Pointer metadata can be extracted from a pointer or reference with the [`metadata`] function.
+/// The data pointer can be extracted by casting a (fat) pointer
+/// to a (thin) pointer to a `Sized` type the `as` operator,
+/// for example `(x: &dyn SomeTrait) as *const SomeTrait as *const ()`.
+///
+/// [dst]: https://doc.rust-lang.org/nomicon/exotic-sizes.html#dynamically-sized-types-dsts
+#[lang = "pointee"]
+pub trait Pointee {
+    /// The type for metadata in pointers and references to `Self`.
+    type Metadata: Copy + Send + Sync + Ord + Hash + Unpin + 'static;
+}
+
+/// Pointers to types implementing this trait alias are “thin”:
+///
+/// ```rust
+/// fn this_never_panics<T: std::ptr::Thin>() {
+///     assert_eq!(std::mem::size_of::<&T>(), std::mem::size_of::<usize>())
+/// }
+/// ```
+pub trait Thin = Pointee<Metadata=()>;
```

Comment:
I believe it's better to actually introduce thin pointers instead of using this trait alias.

The problem is that this trait alias doesn't necessary play well with structs containing DSTs (which might again be "composed" DSTs). It should be possible to have a thin pointer to such a struct and get access to all of it's fields without needing any unsafe code. Except the DST field which would if referenced just return a thin reference to the DST. For many thinks it would make writing unsafe code around all kinds of DST types much easier if we actually had thin pointers. 

For example we could have something like (scatch):

```rust
#[repr(transparent)]
#[lang = "thin_ptr"] //see below
#[fundamental] //maybe?? I forgot what that did
pub struct Thin<T: ?Sized>(T)

//override auto-impl
unsafe impl<T> Pointee for Thin<T> {
    type Metadata = ()
}
```

The reason why it's a lang item is because it has a bit special deref handling, mainly:

- given  `struct Foo { f1: usize, f2: usize, dst: [u8] }`
- given `let foo: &Thin<Foo> = ...`, 
- then `&foo.f1` is of type `&usize`
- then `&foo.f2` is of type `&usize`
- **then `&foo.dst` is of type `&Thin<[u8]>`**

This (or something similar) has some benefits:

- We have a direct representation for thin slices (with unknown length):  `&Thin<[T]>`,`&mut Thin<[T]>`, `*const Thin<[T]>`, `*mut Thin<[T]>`
   - No need to use unsafe "hacks" to get thin pointer
   - This works with all potential future pointer types (e.g. `*raw` or similar) which we maybe might add
- We can make from `from_raw_parts` more type save by accepting `*const Thin<T>` instead of an `*const ()`

By keeping the type instead of using `*const ()` we should need less complex unsafe code to e.g. implement a ThinBox for trait objects shown above we would "just" get the thin pointer to the DST field and then pass it with the metadata to `from_raw_parts`.

Sure this doesn't really eliminate the data alignment fixing done in `data_ptr` it just moves it from the user written library into the implementation of `from_raw_parts` (a `Thin<dyn Trait>` would not be guaranteed to be correctly aligned as we simply can't know the alignment without knowing the type so the `from_raw_parts` impl. would need to do what `data_ptr` does). 

Still moving this tricky bits from user libraries to the core/std library seems to be a good idea I think and the improved type safety around `from_raw_parts` would be good too as (I think) in many case we don't handle completely erased types but just the erasure of the exact metadata used.

Through we probably still would need to have a way to create a `*const Thin<dyn Trait>` from a  `*const ()` for some ffi aspects.

(Note that a `&Thin<dyn Trait>` is kinda pointless but a `&Thin<DSTContainingDynTrait>` isn't as you still could access other fields safely).

Reply:
An additional problem I found is that you can't have unaligned references in rust. It's UB.

So `&Thin<dyn Trait>`  would need to "encapsulate" the unalignedness while not doing so for everything for which we know the alignment i.e. another reason why `Thin` would be a lang item. Keeping with how rust tends to do thinks it might be more appropriate to make `Thin` a "special" syntax. But then we would need to have thinks like `&thin X`, `&mut thin X`, or similar. But having a `Thin<T>` does just work very egonomical with the rest of rust.

(Solved below)

Another problem I found is that `Thin<dyn Trait>` or `Thin<[u8]>` as a unknown layout in the sense that:

- We don't know the size for either.
- We don't know the "internal" alignment/padding of the inner value in `Thin<dyn Trait>` (`Thin<dyn Trait>` is on the outside well but "unknown" aligned), as we can't move it in memory anyway due to now knowing the size we can just report an alignment of 1, but between the start of `Thin<dyn Trait>` and the data of `dyn Trait` there is some unknown padding to "fix" the alignment of the internal field. 

But `Layout::for_value(&Thin<[u8]>)` does still need to return a value. 

So either we return a "not so useful" value like alignment 1 and size of `usize::MAX`, which I don't like or we need to directly have Thin pointers instead of a Thin wrapper around the `T: Pointee` type.

Like `ThinRef<'a,T>`, `ThinMut<'a,T>`, `ThinConstPtr`, `ThinMutPtr`. (as e.g. `#[...] struct ThinMutPtr<T>(ptr: *mut (), PhantomData<T>)`)

Which is slightly less elegant but would still work  otherwise like described for `Thin` above.

This also would *fix the inner alignment problem much easier* as e.g. `ThinRef<'a, T>` could be  implemented as `#[...] struct ThinRef<'a, T> { ptr: *const (), marker: PhantomData<&'a T> }` which works just fine for a not-necessary-correctly aligned `ThinRef<'a, dyn Trait>`. 

The only problem would be if we point to not-allocated memory. But we are guaranteed to point to allocated memory. 
The alignment is only wrong if the pointer is to a DST field, so the only "bad" think is we might point into the padding in front of the field instead of the field directly. Which is fine and expected (see the `data_ptr` example and what I wrote previously about it).

Reply:
> I believe it's better to actually introduce thin pointers instead of using this trait alias.

I don’t understand what "actually introduce" means.

The `Thin` trait alias proposed here describes an existing property of some existing types and the existing pointer/reference types to those. `*const u32` is a thin pointer, `*const str` is not. Therefore `u32: Thin` holds, `str: Thin` does not.

> The problem is that this trait alias doesn't necessary play well with structs containing DSTs (which might again be "composed" DSTs). It should be possible to have a thin pointer to such a struct

In today’s Rust a struct that contains a DST field is itself also a DST and therefore pointers to it are wide, not thin.

Introducing thin pointers *to DSTs* (structs or not) is an entirely new language feature that is not part of this RFC. If you feel that feature should be pursued, consider writing a separate RFC for it.

Reply:
This RFCs would enable a third-party *library* to implement a type like `ThinBox` that stores the metadata in the heap allocation and keeps the `Box`-like pointer thin.

Reply:
We have in the end 2 kinds of DSTs one which I refereed to as composed DST's which have any number of non DST fields and a tailing DST fields and what I will refer to as fundamental DSTs which are currently `[T]` and `dyn Trait`. 

The problem is that without thin pointers to DSTs which allow access to non DST fields any library `ThinBox` implementation would need to either create a fat-pointer or use a lot of unsafe code to access non DST fields.

This makes it harder to implement all kind of custom DST types like e.g.  slice with the length prefixing the slice or similar.
Mainly a lot of unsafe code is needed which with thin pointers to DST can be moved into the `from_raw_parts` implementation in core/std. 

Also semantically the `from_raw_parts` functions(s) does take a thin pointer to an DST and meta data and creates a  potential non-thin pointer.  But instead we represent it as taking a untyped pointer and metadata making the functions more unsafe then they need to be IMHO (as you could put in a "wrong" thinned pointer.

While we seem to clearly disagree on this, thin pointers to DST are IMHO a fundamental part of the API around handling pointer metadata as they are the direct representation of fat-pointers without metadata.

But then a new RFC could also try to amend any non yet stabilized RFC so I guess putting this into a separate RFC is worth a try.

My goal is to long term make it possible to handle/write DST types without needing to know about all kinds of unsafe easy to get wrong without noticing it details like alignment fixing making custom DSTs available to any rust programmer.

I will see if I can find time to write a RFC for this in the next week or so.